// javascript-time-ago/commonjs/style/twitter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.style = void 0;

var _round = _interopRequireDefault(require("../steps/round"));

var _steps = require("../steps");

var _locale = require("../locale");

var _renameLegacyProperties = _interopRequireDefault(require("./renameLegacyProperties"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// For compatibility with the old versions of this library.
// Twitter-style relative date/time formatting.
// ("1m", "2h", "Mar 3", "Apr 4, 2012").
//
// Seconds, minutes or hours are shown for shorter intervals,
// and longer intervals are formatted using full date format.
var steps = [// Seconds
{
  // Format time in seconds.
  formatAs: 'second'
}, // Minutes
{
  // Starts showing `1m` after `59s`.
  minTime: _steps.minute - 0.5,
  formatAs: 'minute'
}, // Hours
{
  // After `59m` it will show `1h`.
  minTime: _steps.hour - 0.5 * _steps.minute,
  formatAs: 'hour'
}]; // A cache for `Intl.DateTimeFormat` formatters
// for various locales (is a global variable).

var formatters = {}; // Starting from day intervals, output month and day.

var monthAndDay = {
  minTime: _steps.day - 0.5 * _steps.hour,
  format: function format(value, locale) {
    /* istanbul ignore else */
    if (!formatters[locale]) {
      formatters[locale] = {};
    }
    /* istanbul ignore else */


    if (!formatters[locale].this_year) {
      // "Apr 11" (MMMd)
      formatters[locale].this_year = new Intl.DateTimeFormat(locale, {
        month: 'short',
        day: 'numeric'
      });
    } // Output month and day.


    return formatters[locale].this_year.format((0, _steps.getDate)(value));
  },
  getTimeToNextUpdate: function getTimeToNextUpdate(date, _ref) {
    var now = _ref.now,
        future = _ref.future;
    var nextYear = new Date(new Date(now).getFullYear() + 1, 0);
    return nextYear.getTime() - now;
  }
}; // If the `date` happened/happens outside of current year,
// then output day, month and year.

var yearMonthAndDay = {
  test: function test(timestamp, _ref2) {
    var now = _ref2.now,
        future = _ref2.future;

    if (future) {
      // Jan 1st 00:00 of the next year.
      var nextYear = new Date(new Date(now).getFullYear() + 1, 0); // If the `date` is future year.

      return timestamp > nextYear.getTime();
    } else {
      // Jan 1st 00:00 of the this year.
      var thisYear = new Date(new Date(now).getFullYear(), 0); // If the `date` is past year.

      return timestamp < thisYear.getTime();
    }
  },
  format: function format(value, locale) {
    /* istanbul ignore if */
    if (!formatters[locale]) {
      formatters[locale] = {};
    }
    /* istanbul ignore else */


    if (!formatters[locale].other) {
      // "Apr 11, 2017" (yMMMd)
      formatters[locale].other = new Intl.DateTimeFormat(locale, {
        year: 'numeric',
        month: 'short',
        day: 'numeric'
      });
    } // Output day, month and year.


    return formatters[locale].other.format((0, _steps.getDate)(value));
  },
  getTimeToNextUpdate: function getTimeToNextUpdate() {
    // Doesn't need to be updated.
    return 1000 * _steps.year;
  }
}; // If `Intl.DateTimeFormat` is supported,
// then longer time intervals will be formatted as dates.

/* istanbul ignore else */

if ((0, _locale.intlDateTimeFormatSupported)()) {
  steps.push(monthAndDay, yearMonthAndDay);
} // Otherwise, if `Intl.DateTimeFormat` is not supported,
// which could be the case when using Internet Explorer,
// then simply mimick "round" steps.
else {
    steps.push({
      minTime: _steps.day - 0.5 * _steps.hour,
      formatAs: 'day'
    }, {
      minTime: _steps.week - 0.5 * _steps.day,
      formatAs: 'week'
    }, {
      minTime: 3.5 * _steps.week,
      formatAs: 'month'
    }, {
      minTime: _steps.year - 0.5 * _steps.month,
      formatAs: 'year'
    });
  }

var style = {
  steps: steps,
  labels: [// "mini-time" labels are only defined for a few languages.
  'mini-time', // "short-time" labels are only defined for a few languages.
  'short-time', // "narrow" and "short" labels are defined for all languages.
  // "narrow" labels can sometimes be weird (like "+5d."),
  // but "short" labels have the " ago" part, so "narrow" seem
  // more appropriate.
  // "short" labels would have been more appropriate if they
  // didn't have the " ago" part, hence the "short-time" above.
  'narrow', // Since "narrow" labels are always present, "short" element
  // of this array can be removed.
  'short'] // For compatibility with the old versions of this library.

};
exports.style = style;

var _default = (0, _renameLegacyProperties.default)(style);

exports.default = _default;
