// javascript-time-ago/commonjs/steps/getStep.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getStep;

var _getStepDenominator = _interopRequireDefault(require("./getStepDenominator"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

/**
 * Finds an appropriate `step` of `steps` for the time interval (in seconds).
 *
 * @param {Object[]} steps - Time formatting steps.
 *
 * @param {number} secondsPassed - Time interval (in seconds).
 *                                 `< 0` for past dates and `> 0` for future dates.
 *
 * @param {number} options.now - Current timestamp.
 *
 * @param {boolean} [options.future] - Whether the date should be formatted as a future one
 *                                     instead of a past one.
 *
 * @param {string[]} [options.units] - A list of allowed time units.
 *                                     (Example: ['second', 'minute', 'hour', …])
 *
 * @param {boolean} [options.getNextStep] - Pass true to return `[step, nextStep]` instead of just `step`.
 *
 * @return {Object|Object[]} [step] — Either a `step` or `[step, nextStep]`.
 */
function getStep(steps, secondsPassed, _ref) {
  var now = _ref.now,
      future = _ref.future,
      units = _ref.units,
      getNextStep = _ref.getNextStep;
  // Ignore steps having not-supported time units in `formatAs`.
  steps = filterStepsByUnits(steps, units);

  var step = _getStep(steps, secondsPassed, {
    now: now,
    future: future
  });

  if (getNextStep) {
    if (step) {
      var nextStep = steps[steps.indexOf(step) + 1];
      return [step, nextStep];
    }

    return [];
  }

  return step;
}

function _getStep(steps, secondsPassed, _ref2) {
  var now = _ref2.now,
      future = _ref2.future;

  // If no steps fit the conditions then return nothing.
  if (steps.length === 0) {
    return;
  } // Find the most appropriate step.


  var i = getStepIndex(steps, secondsPassed, {
    now: now,
    future: future || secondsPassed < 0
  }); // If no step is applicable the return nothing.

  if (i === -1) {
    return;
  }

  var step = steps[i]; // Apply granularity to the time amount
  // (and fall back to the previous step
  //  if the first level of granularity
  //  isn't met by this amount)

  if (step.granularity) {
    // Recalculate the amount of seconds passed based on `granularity`.
    var secondsPassedGranular = Math.round(Math.abs(secondsPassed) / (0, _getStepDenominator.default)(step) / step.granularity) * step.granularity; // If the granularity for this step is too high,
    // then fall back to the previous step.
    // (if there is any previous step)

    if (secondsPassedGranular === 0 && i > 0) {
      return steps[i - 1];
    }
  }

  return step;
}
/**
 * Returns the `minTime` threshold for moving from `fromStep` to `next_step`.
 * @param  {Object} fromStep - From step.
 * @param  {Object} toStep - To step.
 * @param  {number} options.now - Current timestamp.
 * @param  {number} options.secondsPassed - How much seconds have passed since the date till `now`.
 * @param  {boolean} options.future - Whether the time interval should be formatted as a future one.
 * @return {number}
 * @throws Will throw if no `minTime` threshold is found.
 */


function getThresholdForTransition(fromStep, toStep, secondsPassed, options) {
  var threshold;
  var isLegacy; // Allows custom thresholds when moving
  // from a specific step to a specific step.

  if (fromStep) {
    if (_typeof(toStep.minTime) === 'object' && fromStep.id) {
      threshold = toStep.minTime[fromStep.id];
    }

    if (threshold === undefined) {
      // "threshold_for_idOrUnit: value" is a legacy way of specifying "minTime: { id: value }".
      // Developers should use "minTime" property instead of "threshold".
      if (fromStep.id || fromStep.unit) {
        threshold = toStep["threshold_for_".concat(fromStep.id || fromStep.unit)];

        if (threshold !== undefined) {
          isLegacy = true;
        }
      }
    }
  } // If no custom `minTime` threshold is defined for this "fromStep -> toStep" transition
  // then use the default `minTime` threshold for the next step.


  if (threshold === undefined) {
    // "threshold" is a legacy name of "minTime".
    if (toStep.threshold !== undefined) {
      threshold = toStep.threshold;
      isLegacy = true;
    } else {
      if (_typeof(toStep.minTime) === 'object') {
        threshold = toStep.minTime.default;
      } else {
        threshold = toStep.minTime;
      }
    }
  } // Evaluate the `test()` function.


  if (threshold === undefined) {
    if (toStep.test) {
      var now = options.now,
          future = options.future;

      if (toStep.test(now - secondsPassed * 1000, {
        now: now,
        future: future
      })) {
        // `0` threshold always passes.
        threshold = 0;
      } else {
        // `MAX_SAFE_INTEGER` threshold won't ever pass in real life.
        threshold = 9007199254740991; // Number.MAX_SAFE_INTEGER
      }
    }
  } // Convert the threshold to a number.


  if (typeof threshold === 'function') {
    var _now = options.now,
        _future = options.future; // if (isLegacy) {

    threshold = threshold(_now, _future); // } else {
    // 	threshold = threshold(secondsPassed, { now, future })
    // }
  } // Throw if no `minTime` threshold is found.


  if (fromStep && typeof threshold !== 'number') {
    // Babel transforms `typeof` into some "branches"
    // so istanbul will show this as "branch not covered".

    /* istanbul ignore next */
    var type = _typeof(threshold);

    throw new Error("[javascript-time-ago] Each step must define either `minTime` or `test()`, except for the first one. Got \"".concat(threshold, "\", ").concat(type, ". Step: ").concat(JSON.stringify(toStep)));
  }

  return threshold;
}
/**
 * Iterates through steps until it finds the maximum one satisfying the `minTime` threshold.
 * @param  {Object} steps - Steps.
 * @param  {number} secondsPassed - How much seconds have passed since the date till `now`.
 * @param  {number} options.now - Current timestamp.
 * @param  {boolean} options.future - Whether the time interval should be formatted as a future one.
 * @param  {number} [i] - Gradation step currently being tested.
 * @return {number} Gradation step index.
 */


function getStepIndex(steps, secondsPassed, options) {
  var i = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : 0;

  // If the `minTime` threshold for moving from previous step
  // to this step is too high then return the previous step.
  if (Math.abs(secondsPassed) < getThresholdForTransition(steps[i - 1], steps[i], secondsPassed, options)) {
    return i - 1;
  } // If it's the last step then return it.


  if (i === steps.length - 1) {
    return i;
  } // Move to the next step.


  return getStepIndex(steps, secondsPassed, options, i + 1);
}
/**
 * Leaves only allowed steps.
 * @param  {Object[]} steps
 * @param  {string[]} units - Allowed time units.
 * @return {Object[]}
 */


function filterStepsByUnits(steps, units) {
  return steps.filter(function (_ref3) {
    var unit = _ref3.unit,
        formatAs = _ref3.formatAs;
    // "unit" is now called "formatAs".
    unit = unit || formatAs; // If this step has a `unit` defined
    // then this `unit` must be in the list of allowed `units`.

    if (unit) {
      return units.indexOf(unit) >= 0;
    } // A step is not required to specify a `unit`:
    // alternatively, it could specify `format()`.
    // (see "twitter" style for an example)


    return true;
  });
}
